File: backend/app/schemas/email.py
from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict, EmailStr, Field

from app.models.email import EMAIL_FOLDERS, EMAIL_PRIORITIES, EMAIL_STATUSES, TEMPLATE_CATEGORIES


class EmailRecipientInput(BaseModel):
    recipient_type: str = Field(pattern="^(to|cc|bcc)$")
    email_address: EmailStr
    display_name: str | None = None
    user_id: UUID | None = None
    contact_id: UUID | None = None


class EmailRecipientResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    recipient_type: str
    email_address: str
    display_name: str | None
    user_id: UUID | None
    contact_id: UUID | None


class EmailAttachmentResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    filename: str
    content_type: str
    size_bytes: int
    created_at: datetime


class EmailLogResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    event_type: str
    metadata: dict | None = None
    created_at: datetime


class UserRef(BaseModel):
    id: UUID
    full_name: str
    email: str


class EmailCreate(BaseModel):
    subject: str = Field(default="", max_length=500)
    body_html: str | None = None
    body_text: str | None = None
    priority: str = Field(default="normal", pattern="^(low|normal|high|urgent)$")
    recipients: list[EmailRecipientInput] = Field(default_factory=list)
    company_id: UUID | None = None
    contact_id: UUID | None = None
    lead_id: UUID | None = None
    deal_id: UUID | None = None
    task_id: UUID | None = None
    meeting_id: UUID | None = None
    template_id: UUID | None = None
    parent_email_id: UUID | None = None
    thread_id: UUID | None = None
    from_name: str | None = None
    metadata: dict | None = None


class EmailUpdate(BaseModel):
    subject: str | None = Field(default=None, max_length=500)
    body_html: str | None = None
    body_text: str | None = None
    priority: str | None = Field(default=None, pattern="^(low|normal|high|urgent)$")
    recipients: list[EmailRecipientInput] | None = None
    company_id: UUID | None = None
    contact_id: UUID | None = None
    lead_id: UUID | None = None
    deal_id: UUID | None = None
    task_id: UUID | None = None
    meeting_id: UUID | None = None
    is_read: bool | None = None
    is_starred: bool | None = None
    is_important: bool | None = None
    from_name: str | None = None
    metadata: dict | None = None


class EmailSendRequest(BaseModel):
    email_id: UUID | None = None
    subject: str = Field(default="", max_length=500)
    body_html: str | None = None
    body_text: str | None = None
    priority: str = Field(default="normal", pattern="^(low|normal|high|urgent)$")
    recipients: list[EmailRecipientInput] = Field(min_length=1)
    company_id: UUID | None = None
    contact_id: UUID | None = None
    lead_id: UUID | None = None
    deal_id: UUID | None = None
    task_id: UUID | None = None
    meeting_id: UUID | None = None
    template_id: UUID | None = None
    parent_email_id: UUID | None = None
    from_name: str | None = None
    include_signature: bool = True


class EmailScheduleRequest(EmailSendRequest):
    scheduled_at: datetime


class EmailReplyRequest(BaseModel):
    body_html: str | None = None
    body_text: str | None = None
    recipients: list[EmailRecipientInput] | None = None
    include_signature: bool = True
    reply_all: bool = False


class EmailForwardRequest(BaseModel):
    body_html: str | None = None
    body_text: str | None = None
    recipients: list[EmailRecipientInput] = Field(min_length=1)
    include_signature: bool = True


class EmailResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    tenant_id: UUID
    thread_id: UUID | None
    parent_email_id: UUID | None
    sender_id: UUID | None
    sender: UserRef | None = None
    created_by_id: UUID | None
    created_by: UserRef | None = None
    template_id: UUID | None
    from_email: str | None
    from_name: str | None
    subject: str
    body_html: str | None
    body_text: str | None
    status: str
    priority: str
    folder: str
    direction: str
    is_read: bool
    is_starred: bool
    is_important: bool
    has_attachments: bool
    scheduled_at: datetime | None
    sent_at: datetime | None
    archived_at: datetime | None
    company_id: UUID | None
    contact_id: UUID | None
    lead_id: UUID | None
    deal_id: UUID | None
    task_id: UUID | None
    meeting_id: UUID | None
    activity_id: UUID | None
    metadata: dict | None = None
    recipients: list[EmailRecipientResponse] = []
    attachments: list[EmailAttachmentResponse] = []
    logs: list[EmailLogResponse] = []
    created_at: datetime
    updated_at: datetime


class EmailListResponse(BaseModel):
    items: list[EmailResponse]
    total: int
    page: int
    page_size: int
    total_pages: int


class EmailTemplateCreate(BaseModel):
    name: str = Field(min_length=1, max_length=255)
    category: str = Field(default="sales")
    subject: str = Field(min_length=1, max_length=500)
    body_html: str = Field(min_length=1)
    body_text: str | None = None
    variables: list[str] | None = None


class EmailTemplateUpdate(BaseModel):
    name: str | None = Field(default=None, max_length=255)
    category: str | None = None
    subject: str | None = Field(default=None, max_length=500)
    body_html: str | None = None
    body_text: str | None = None
    variables: list[str] | None = None


class EmailTemplateResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    tenant_id: UUID
    name: str
    category: str
    subject: str
    body_html: str
    body_text: str | None
    variables: list[str] | None = None
    created_by_id: UUID | None
    created_by: UserRef | None = None
    created_at: datetime
    updated_at: datetime


class EmailTemplateListResponse(BaseModel):
    items: list[EmailTemplateResponse]
    total: int


class EmailStatisticsResponse(BaseModel):
    unread_count: int
    drafts_count: int
    scheduled_count: int
    sent_today: int
    sent_this_week: int
    open_rate: float
    click_rate: float
    reply_rate: float
    delivery_rate: float


class EmailUserSettingsResponse(BaseModel):
    signature_html: str | None
    signature_text: str | None
    default_from_name: str | None


class EmailUserSettingsUpdate(BaseModel):
    signature_html: str | None = None
    signature_text: str | None = None
    default_from_name: str | None = None


class EmailThreadResponse(BaseModel):
    thread_id: UUID
    subject: str
    last_message_at: datetime | None
    emails: list[EmailResponse]

File: frontend/types/email.ts
export const EMAIL_FOLDERS = [
  "inbox",
  "sent",
  "drafts",
  "scheduled",
  "starred",
  "archive",
  "trash",
] as const;

export type EmailFolder = (typeof EMAIL_FOLDERS)[number];

export const EMAIL_PRIORITIES = ["low", "normal", "high", "urgent"] as const;
export type EmailPriority = (typeof EMAIL_PRIORITIES)[number];

export const EMAIL_STATUSES = ["draft", "scheduled", "sending", "sent", "failed", "received"] as const;
export type EmailStatus = (typeof EMAIL_STATUSES)[number];

export const TEMPLATE_CATEGORIES = [
  "sales",
  "support",
  "follow_up",
  "welcome",
  "reminder",
  "marketing",
] as const;

export type TemplateCategory = (typeof TEMPLATE_CATEGORIES)[number];

export interface EmailRecipient {
  id?: string;
  recipient_type: "to" | "cc" | "bcc";
  email_address: string;
  display_name?: string | null;
  user_id?: string | null;
  contact_id?: string | null;
}

export interface EmailAttachment {
  id: string;
  filename: string;
  content_type: string;
  size_bytes: number;
  created_at: string;
}

export interface EmailLog {
  id: string;
  event_type: string;
  metadata?: Record<string, unknown> | null;
  created_at: string;
}

export interface EmailUserRef {
  id: string;
  full_name: string;
  email: string;
}

export interface Email {
  id: string;
  tenant_id: string;
  thread_id?: string | null;
  parent_email_id?: string | null;
  sender_id?: string | null;
  sender?: EmailUserRef | null;
  created_by_id?: string | null;
  created_by?: EmailUserRef | null;
  template_id?: string | null;
  from_email?: string | null;
  from_name?: string | null;
  subject: string;
  body_html?: string | null;
  body_text?: string | null;
  status: EmailStatus;
  priority: EmailPriority;
  folder: EmailFolder | string;
  direction: string;
  is_read: boolean;
  is_starred: boolean;
  is_important: boolean;
  has_attachments: boolean;
  scheduled_at?: string | null;
  sent_at?: string | null;
  archived_at?: string | null;
  company_id?: string | null;
  contact_id?: string | null;
  lead_id?: string | null;
  deal_id?: string | null;
  task_id?: string | null;
  meeting_id?: string | null;
  activity_id?: string | null;
  metadata?: Record<string, unknown> | null;
  recipients: EmailRecipient[];
  attachments: EmailAttachment[];
  logs: EmailLog[];
  created_at: string;
  updated_at: string;
}

export interface EmailListResponse {
  items: Email[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
}

export interface EmailTemplate {
  id: string;
  tenant_id: string;
  name: string;
  category: TemplateCategory | string;
  subject: string;
  body_html: string;
  body_text?: string | null;
  variables?: string[] | null;
  created_by_id?: string | null;
  created_by?: EmailUserRef | null;
  created_at: string;
  updated_at: string;
}

export interface EmailStatistics {
  unread_count: number;
  drafts_count: number;
  scheduled_count: number;
  sent_today: number;
  sent_this_week: number;
  open_rate: number;
  click_rate: number;
  reply_rate: number;
  delivery_rate: number;
}

export interface EmailUserSettings {
  signature_html?: string | null;
  signature_text?: string | null;
  default_from_name?: string | null;
}

export interface EmailFilters {
  folder?: EmailFolder | string;
  q?: string;
  unread?: boolean;
  starred?: boolean;
  important?: boolean;
  has_attachments?: boolean;
  scheduled?: boolean;
  company_id?: string;
  contact_id?: string;
  lead_id?: string;
  deal_id?: string;
  entity_type?: string;
  entity_id?: string;
  page?: number;
  page_size?: number;
  sort_by?: string;
  sort_order?: "asc" | "desc";
}

export interface EmailInput {
  subject?: string;
  body_html?: string;
  body_text?: string;
  priority?: EmailPriority;
  recipients?: EmailRecipient[];
  company_id?: string | null;
  contact_id?: string | null;
  lead_id?: string | null;
  deal_id?: string | null;
  task_id?: string | null;
  meeting_id?: string | null;
  template_id?: string | null;
  parent_email_id?: string | null;
  from_name?: string | null;
}

export interface EmailSendInput extends EmailInput {
  email_id?: string | null;
  include_signature?: boolean;
  recipients: EmailRecipient[];
}

export interface EmailScheduleInput extends EmailSendInput {
  scheduled_at: string;
}
